feat: Add baggage span processor

MikeGoldsmith opened the pull request:
### Summary of changes

Adds the baggage span processor. I've added under new root dir `processors`, not sure if that's the best place for it or it should go somewhere else. Happy to take direction.

- Closes #934

robbkidd commented on processor/baggage/lib/opentelemetry/processor/baggage/baggage_span_processor.rb:
```diff
@@ -0,0 +1,79 @@
+      # @example
+      #   OpenTelemetry::SDK.configure do |c|
+      #     # Add the BaggageSpanProcessor to the collection of span processors
+      #     c.add_span_processor(Honeycomb::OpenTelemetry::Trace::BaggageSpanProcessor.new)
```
```suggestion
      #     c.add_span_processor(OpenTelemetry::Processor::Baggage::BaggageSpanProcessor.new)
```

... depending on how the folder structure and module namespace for this processor shakes out her in contrib.

robbkidd reviewed (changes requested):
Here's the first take on the long-awaited span processor to take current Baggage key/values and add them as attributes to spans as they start.

Not sure of the module namespace in the current state. This is a span processor, so should we nest it more deeply under the OpenTelemetry::Trace namespace? Is "Span" in its name sufficient? What do y'all think?

We can work on adding unit tests to this. The first edition of those would look a lot like [the RSpec specs I wrote over here](https://github.com/honeycombio/honeycomb-opentelemetry-ruby/pull/5/files#diff-e855dde1110555a48bae7b2a24913aa3fe1021025982e00e1df75fa01de20bd2).

arielvalentin commented:
Something that I want to confirm, this is still going to require that the user annotate the active/current span _in addition_ to using baggage. 

It's also going to transfer these attributes across service boundaries. Is that the effect that you want?

If so, I think it is important to highlight that this will potentially cause message and HTTP headers to impact performance since there isn't anything built in to say prevent the headers from growing unwieldy. (No per-header compression e.g.) 

In addition to that baggage usage means library instrumentation will include all items in outbound headers to 3rd parties, which will raise concerns for privacy.

I realize that these concerns I am highlighting are not specific to this implementation but issues with Baggage; however, I think it would be important to highlight these concerns so that Baggage users are not caught off guard.

```ruby
          span.add_attributes('user_id' => user.id, 'cart_id' => cart.id)
          ctx = OpenTelemetry::Baggage.build do |builder|

            builder.set_value('user_id', user.id)
            builder.set_value('cart_id', cart.id)
          end

          tracer.in_span('checkout') do |span|
              # yada yada yada
          end 
```

MikeGoldsmith commented:
> Something that I want to confirm, this is still going to require that the user annotate the active/current span _in addition_ to using baggage.
> 
> It's also going to transfer these attributes across service boundaries. Is that the effect that you want?
> 
> If so, I think it is important to highlight that this will potentially cause message and HTTP headers to impact performance since there isn't anything built in to say prevent the headers from growing unwieldy. (No per-header compression e.g.)
> 
> In addition to that baggage usage means library instrumentation will include all items in outbound headers to 3rd parties, which will raise concerns for privacy.
> 
> I realize that these concerns I am highlighting are not specific to this implementation but issues with Baggage; however, I think it would be important to highlight these concerns so that Baggage users are not caught off guard.
> 
> ```ruby
>           span.add_attributes('user_id' => user.id, 'cart_id' => cart.id)
>           ctx = OpenTelemetry::Baggage.build do |builder|
> 
>             builder.set_value('user_id', user.id)
>             builder.set_value('cart_id', cart.id)
>           end
> 
>           tracer.in_span('checkout') do |span|
>               # yada yada yada
>           end 
> ```

Yes, it is known that baggage entries are transported to 3rd parties and you're correct this is a fault of baggage not this processor in particular. I haven't added a README yet, but will add warnings there, in the class implementation and any module functions to make it as clear as possible. I'll add a note regarding unbounded header size too.

ericmustin commented:
Do any other languages have an implementation in contrib, or more broadly, do any other languages have non-spec processors that live in contrib?

(Fwiw I'm fine with this PR just kinda curious if y'all are setting precedent here, which, again, is fine if so)

robbkidd commented:
> Do any other languages have an implementation in contrib, or more broadly, do any other languages have non-spec processors that live in contrib?

We (Honeycomb) have started opening PRs to add this to contrib in other languages, so: technically "no" but also maybe "yes" soon.

ericmustin commented:
Huzzah. Ty @robbkidd , Sounds good to me. Ping if this gets stuck in the mud happy to help w/approvals but otherwise will leave to the folks more tuned in with uh..the world (I'm on pto for a little while)

robbkidd commented:
Well, I succeeded in adding [a new set of jobs for Processors](https://github.com/open-telemetry/opentelemetry-ruby-contrib/actions/runs/8787708401) to the CI matrix! 🙌🏻 

Now to wait for the matrix ... ⏳

robbkidd commented:
Horrah! [BaggageSpanProcessor tests pass](https://github.com/open-telemetry/opentelemetry-ruby-contrib/actions/runs/8791124624/job/24124658730). 🎉

arielvalentin commented on processor/baggage/lib/opentelemetry/processor/baggage/baggage_span_processor.rb:
💭 Stand-alone vs delegators (decorators)

I mentioned this to @robbkidd during the SIG but if you want to guarantee processing in order you may want to consider wrapping the Simple or Batch Span Processor then use a simple delegator to forward any un-implemented methods allowing this processor to work with any future API changes. 

I tried finding guidance in the specification but did not see it in the latest version. 

https://github.com/open-telemetry/opentelemetry-specification/issues/2459

If that does not appeal to you, then you may also want to consider inheriting from `OpenTelemetry::SDK::Trace::SpanProcessor` and override only the methods that are of interest to you.

MikeGoldsmith replied:
Updated to inherit `SpanProcessor` in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/937/commits/8877e8436721f71a8e440dde8447b7695e0cc400

arielvalentin reviewed (changes requested):
My only request for an actual change is to consolidate the CI jobs. 

The other comments are optional changes/food for thought.

MikeGoldsmith commented:
Added this follow-up enhancement issue to track adding a key predicate function to control what keys are copied to newly created span attributes.
- https://github.com/open-telemetry/opentelemetry-ruby-contrib/issues/956

MikeGoldsmith commented on processor/baggage/lib/opentelemetry/processor/baggage/baggage_span_processor.rb:
```diff
@@ -0,0 +1,85 @@
+module OpenTelemetry
+  module Processor
+    module Baggage
+      # The BaggageSpanProcessor reads key/values stored in Baggage in the
```
Added to README in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/937/commits/9fef03556c4b2437cd10378939bf6dd82299dd9a

arielvalentin commented on CODEOWNERS:
```diff
@@ -31,3 +31,5 @@ instrumentation/racecar/ @chrisholmes @open-telemetry/ruby-contrib-maintainers @
 
 instrumentation/que/ @indrekj @open-telemetry/ruby-contrib-maintainers @open-telemetry/ruby-contrib-approvers @fbogsany @mwear @robertlaurin @dazuma @ericmustin @arielvalentin @ahayworth @plantfansam @robbkidd @simi @kaylareopelle @xuan-cao-swi
+
+processor/baggage/ @robbkidd @mikegoldsmith @open-telemetry/ruby-contrib-maintainers @open-telemetry/ruby-contrib-approvers @fbogsany @mwear @robertlaurin @dazuma @ericmustin @arielvalentin @ahayworth @plantfansam @robbkidd @simi @kaylareopelle @xuan-cao-swi
```
Hmmm... I'll have to take a closer look at our settings then and see why this is failing.

dmathieu replied:
The CODEOWNERS file requires folks to have write access to a repo, hence the errors.

arielvalentin replied:
Of course. As a GitHub employee I should know this.

MikeGoldsmith commented on .github/workflows/ci-processors.yml:
```diff
@@ -0,0 +1,62 @@
+name: CI Processors
```
Combined processors workflow into contrib workflow in https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/937/commits/d053d1ef0a6029970ce58168f4675d31af9c8d71

MikeGoldsmith commented:
> My only request for an actual change is to consolidate the CI jobs.
> 
> The other comments are optional changes/food for thought.

Thanks for your feedback @arielvalentin. I've worked through and addressed your comments. Let me know if there's anything else.

arielvalentin reviewed (approved):
Thank you for your contribution!

arielvalentin commented on processor/baggage/test/opentelemetry/processor/baggage/baggage_span_processor_test.rb:
```diff
@@ -0,0 +1,95 @@
+
+  describe '#on_start' do
+    it 'adds current baggage keys/values as attributes when a span starts' do
+      span.expect(:add_attributes, span, [{ 'a_key' => 'a_value' }])
```
It was a suggestion but really, I leave the final decision up to you.

MikeGoldsmith commented:
@arielvalentin - I've pulled in the change jruby 9.4.6.0 update and updated the new processors workflow. Is there anything else you need / want before accepting this PR?
